### DISCUSSION TITLE
Allow configuration of temperature entities' units

### DISCUSSION THREAD

**Discussion by GrimmiMeloni** [OStrama/weishaupt_modbus#49] [Ideas]
Hey,
I am currently affected by https://github.com/home-assistant/core/issues/120294
TL;DR - on every restart several (not all?) sensors in my HA flip back to Fahrenheit - including this Weishaupt one. This is super annyoing, as it cannot be fixed properly by switching back to Celsius. Any data recorded while the system was in Fahrenheit mode is borked for ever.

While I _**do**_ think think this is an issue in HA core, I am wondering if this could be worked around in the Weishaupt addon?
I don't know anything about homeassistant add-on development, so please ignore my ignorance. 
Is there maybe a way for add-ons to override the cental temperature unit configuration? I know it can be done for entities, each one by one, but that's super tedious. Maybe there is a chance for an easier more central solution?

**Comment by MadOne** [comment]
Hi,

from homeassistant.const import UnitOfTemperature
TEMPERATUR = UnitOfTemperature.CELSIUS

That is what we use. HomeAsisstant can convert these units. 
I dont think we can affect the global unit configuration.
Sadly i dont see any solution that would help you.

If you or anyone else has new info / solutions please share.

**Reply by GrimmiMeloni** [reply]
Too bad. But thanks a lot for checking. Appreciated.

**Comment by OStrama** [comment]
How is this parametrized in your system:
![{0FE70F94-709D-4734-9088-1780FDF22D66}](https://github.com/user-attachments/assets/83814f56-f0ee-427b-88b8-3449ad750f20)

**Reply by GrimmiMeloni** [reply]
Mine is set like yours - even if the sensors report Fahrenheit after the restart.

I just had another restart (due to docker host restarting over night).
Since I have now put together a few dashboards, I can definitely confirm it is not all sensors (or integrations) doing the flip to Fahrenheit. My Homematic integration for example keeps reporting in Celsius.

**Reply by GrimmiMeloni** [reply]
OK, more experimentation on my end.
I tried to rule out the autosensing of HA, and have now set the system's `unit_system` via the `configuration.yaml` instead of the UI.
Interesting enough - my homematic sensors follow the setting, i.e. I deliberately set `us_customary` and they followed along. Same then for setting to `metric`. However, the weishaupt sensors don't care and come always up with fahrenheit.

**Reply by MadOne** [reply]
I had a short look at the homematic integration and i did not see any difference on how they define the temperatures. 
But this is at least a hint where we can investigate further.

**Reply by MadOne** [reply]
Can you please write a bug report with all info you got?

**Reply by GrimmiMeloni** [reply]
Yeah, though not sure what to include.

But here's a hunch: I don't know how (multiple?) inheritance works with Python (sorry, GoLang/Java here), but is it possible that `MyEntity` does actually not have a field `_attr_native_unit_of_measurement`?  I am asking as to my understanding there is no direct inheritance relationship between `MyEntity` and `SensorEntity` where this would actually be defined. Same seems to be true for other `_attr_*` fields, which are set during `__init__` of `MyEntity`, but are then never touched again.
(specifically referring to `entities.py` line 84 ff)

 
To me this signals there's an intention of inheritance here, and maybe that is not working as expected?

**Reply by OStrama** [reply]
Regarding the inheritance I had the same concerns. (It‘s my 1st Python project coming from C/C++.) I think it works as the mentioned variables are part of the Entity class, which is the base class of MyEntity and SensorEntity etc.
However, I tested it with other units, on my 2 system these are definitely used.

**Reply by GrimmiMeloni** [reply]
Dug a bit further into the whole inheritence thing, as I was not satisfied with your explanation. Main reason is that Entity does NOT have that said variable/member called `_attr_native_unit_of_measurement`. Based on what I learned, the code is still correct, though.

What happens is that `SensorEntity` defines a protected attribute `_attr_native_unit_of_measurement`, but does not initialize it. Hence it also does not become part of the dictionary (see `__dict__`) and plainly does not exist. Now what really breaks my poor Java OO brain is that this works by just having some part of the code (in this case the initializer of the hierarchically unrelated `MyEntity`) just define this attribute. That's when it becomes part of the dictionary for the instance of `MySensorEntity`.
`SensorEntity` guards its internal access to `_attr_native_unit_of_measurement` with `hasattr(self, '_attr_native_unit_of_measurement')`, so this does not break if the attribute was never set by any class in the hierarchy.
Oh boy.... today I learned. 😅 

OK, so how can I best drive this forward? What do you folks need from me to chase this further? Can I enable some debug logs? Or is there even a way I could dump the internal states, beyond what the developer view in the HA UI offers?

**Comment by OStrama** [comment]
Interesting! Thanks for taking the time to dig into it!
Meanwhile I made some changes in handling entity attributes. They now are set in the PARAMS_* dictionaries in hpconst.py.
Could you please install the last 1.0.5pre3? It would be interesting to know what happens on your installation if you change the „unit“ member to something different than a temperature unit. This could help to find out if the behavior depends from the deviceclass or the unit..

**Reply by MadOne** [reply]
Perhaps we should try to initialize the MySensorEntity with _attr_native_unit_of_measurement instead of inheriting from MyEntity.

**Comment by MadOne** [comment]
Entity has     _attr_unit_of_measurement: str | None

whats the difference between nativ_unit_of measurement and unit_of_measurement?

**Reply by GrimmiMeloni** [reply]
Unclear. I got this from the definition of the [Sensor Entity in the API docs](https://developers.home-assistant.io/docs/core/entity/sensor).

**Reply by MadOne** [reply]
The unit of measurement that the sensor's value is expressed in. If the native_unit_of_measurement is °C or °F, and its device_class is temperature, the sensor's unit_of_measurement will be the preferred temperature unit configured by the user and the sensor's state will be the native_value after an optional unit conversion. If a unit translation is provided, native_unit_of_measurement should not be defined.

**Reply by MadOne** [reply]
Entity options​

Sensors can be configured by the user, this is done by storing sensor entity options in the sensor's entity registry entry.

OptionDescription:
unit_of_measurementThe sensor's unit of measurement can be overridden for sensors with device class SensorDeviceClass.PRESSURE or SensorDeviceClass.TEMPERATURE.

**Reply by MadOne** [reply]
I think we are getting closer

**Reply by OStrama** [reply]
I made a test with using this:
self._attr_unit_of_measurement = self._api_item.params.get("unit", "")
Result: I do not see any unit anymore...

**Comment by GrimmiMeloni** [comment]
Just another update. I just deliberately restarted HA to inspect the entities post reboot. They end up with unit of measure as Fahreheit.

<img width="1044" alt="image" src="https://github.com/user-attachments/assets/29a20cdc-b607-4190-9934-5a908214adcb">

I will try the patched Version that @OStrama mentioned in a moment and check if it makes a difference.

**Reply by GrimmiMeloni** [reply]
"in a moment" just turns out in to a [little nightmare](https://github.com/OStrama/weishaupt_modbus/discussions/73). Sigh...
This will probably take a bit longer than expected. Need to solve #73 first.

**Comment by OStrama** [comment]
Sorry for that, please see comment in the other thread..

**Comment by GrimmiMeloni** [comment]
OK, seems I really messed up here. 😢 

I made the jump from 0.0.8 to 1.0.4 and the problem has disappeared. The Celsius settings do remain now also through restarts. I just confirmed this. 

I am sorry for having wasted everyone's time on this issue.

**Reply by MadOne** [reply]
Glad it works now as intended. 0.0.8 is really old. Hope we get soon merged into the hacs repo. So no more custom repositorys are required.
